routing: Fix route collisions in AWS ENI

christarazi opened the pull request:
See commit msgs.

This PR has been "forward-ported" from the following direct backport PR to the v1.7 branch: https://github.com/cilium/cilium/pull/14337.

Most commits have been forward-ported from the aforementioned PR, but a few commits are only meant for master (with intention to be backported to v1.9 & 1.8) because of code that didn't exist in v1.7 and because of the initial assumption that this issue impacts both ENI and Azure modes. In reality, only ENI mode has been impacted so far. While it is possible for the issue to occur on Azure, it is very unlikely and a separate PR will be made to address that (see https://github.com/cilium/cilium/issues/14705).

Fixes: https://github.com/cilium/cilium/issues/14336

christarazi committed:
routing: Fix route collisions in AWS ENI

[ upstream commit 332a3fd41; forward-ported from v1.7 tree ]

This commit fixes a potential route collision in AWS ENI IPAM modes,
where the ifindex could equal the main routing table ID (from 253-255)
[1], causing traffic to be subject to these routes incorrectly. This is
admittedly rare, but we've seen this from a user report. The impact is
that most traffic on the node is suddenly blackholed.

To fix this, we say that each device or interface (ENI) will have their
own dedicated routing table. The table ID will start with an offset of
10 because it is highly unlikely to collide with the main routing table
ID (from 253-255). We grab the number associated with the ENI device
(`Number`) and add the offset. For example, if we have an ENI device
"eni-0" which has a `Number` of 5, then the table ID will be 10 + 5 =
15.

Another important piece to note is that only the egress rule will reside
inside the per-device tables, whereas the ingress rule will stay in the
main routing table. This is because we want the main routing table to
hold the routes to the endpoint.

Moving forward, the ENI datapath will now create rules under a new
egress priority value (RulePriorityEgressv2), as long as the
egress-multi-home-ip-rule-compat flag is false. If it's true, then the
datapath will create rules under the original egress priority value
(RulePriorityEgress). This helps disambiguate when running with the
older or newer ENI datapath.

See https://github.com/cilium/cilium/issues/14336.

[1]: See ip-route(8)

Reported-by: Vlad Ungureanu <vladu@palantir.com>
Suggested-by: Joe Stringer <joe@cilium.io>
Suggested-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -49,17 +51,29 @@ func (info *RoutingInfo) Configure(ip net.IP, mtu int, masq bool) error {
 
-	// Route all traffic to the interface address via the main routing table
+	// Route all traffic to the interface address via the main routing table.
+	// Ingress rules are an exception to the above comment.
```
I can't find the historical context on why ingress rules need to be in the main routing table. 

Side note, if we decide to change it to be the same as the egress rule, then we'd have to change the protoype of `Delete()` which only accepts an IP. We wouldn't have enough information to find out which table the ingress rule is in because we'd need the iface name to retrieve the table ID. Just wanted to throw this out there in case we go down this route.

christarazi commented on pkg/datapath/linux/routing/routing.go:
```diff
@@ -192,6 +208,29 @@ func deleteRule(r route.Rule) error {
+
+	// We then use the interface name to get the integer part of the iface name
+	// (e.g. "0" from eth0, "1" from eth1).
+	idx, err := strconv.ParseInt(strings.ReplaceAll(iface.Attrs().Name, "eth", ""), 10, 64)
```
Are Azure interfaces always "eth*"? I assumed this just to get something out to get feedback. This is a minor detail that can be iterated on later.

ungureanuvladvictor replied:
I don't think the eth replace will work because the naming of the interfaces is different if the user enables or not (https://en.wikipedia.org/wiki/Consistent_Network_Device_Naming).

If this is disabled then I think the behavior is to keep `eth*` but if it's enabled it can differ (stock ubuntu AMI names interfaces `ens*`).

joestringer replied:
Yeah, here's an example Ubuntu device I have on my laptop: `enp0s31f6`. These numbers are derived from the hardware layout of the system so that they would be consistent regardless of when each device is initialized during kernel boot.

christarazi replied:
How can we reliably extract the "relevant" part from the ifname if don't know what the format is and is unpredictable? 

My initial thought was to grab the last digits of the ifname, essentially a regex of `[0-9]+$`. Would that work?

christarazi committed:
daemon, option: Add flag egress-multi-home-ip-rule-compat

[ upstream commit 549c256f9; forward-ported from v1.7 tree ]

This flag is needed to control the behavior of Cilium when it starts up
under ENI mode.

If the flag is false, meaning "do not maintain compatibility", then
Cilium will attempt to migrate the ENI datapath (`ip rule`s and routes)
created under the aforementioned IPAM mode to a new table ID scheme.
The table ID refers to the Linux routing policy database tables, aka
"routing table".

If the flag is true, meaning "maintain compatibility", then Cilium will
not attempt to migrate the ENI datapath under the aforementioned IPAM
mode to the new table ID scheme. It will continue to use the original
scheme. Additionally, when the flag is true and Cilium finds the rules
under the newer scheme (by checking the priority of the rule), it will
attempt to migrate back to the original scheme. This allows downgrading
Cilium without affecting connectivity.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
routing: Add ENI route table migration logic

[ upstream commit 1b9ed4452; forward-ported from v1.7 tree ]

This commit will fixup the ENI datapath depending on the
egress-multi-home-ip-rule-compat flag (see previous commits for
context). The migration logic supports both upgrading and downgrading
the ENI datapath.

This logic must run on startup before the API is served and before the
health endpoint is created, so that no endpoints are prematurely
crreated before Cilium has had the chance to migrate the entire
datapath.

See https://github.com/cilium/cilium/issues/14336.

Suggested-by: Joe Stringer <joe@cilium.io>
Suggested-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
api: Extend IPAM to accept interface number

[ upstream commit a5d696a4e; forward-ported from v1.7 tree ]

This is needed in ENI mode. In upcoming commits, the interface number
(ENI.Number) will be used to compute the per-ENI table ID in order to
store rules and routes for the ENI datapath.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
cni, routing: Plumb interface number

[ upstream commit b86bf9368; forward-ported from v1.7 tree ]

In the previous commit, we added the interface number to the IPAM
response for ENI mode. This commit plumbs this new field into the CNI to
set up the ENI datapath.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
linux_defaults: Add RulePriorityEgressv2

[ upstream commit aec2d8f24; forward-ported from v1.7 tree ]

This new priority value is vital for disambiguating which rules are
still under the old scheme. Without this, upgrading to the new scheme
would be difficult, as we aren't able to identify which rules have been
fixed up [1]. Furthermore, this would also allow us to enable downgrades
from the new scheme, because we would be able to identify which rules
need to be modified.

[1]: https://github.com/cilium/cilium/issues/14336

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
revert: Add ability to extend the revert stack

[ upstream commit db23c506f; forward-ported from v1.7 tree ]

This is useful to aggregate the items to revert in one stack, so that it
can all be done at once.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
linux_defaults: Add RouteTableInterfacesOffset

[ upstream commit e0525602b; forward-ported from v1.7 tree ]

This new value is the table ID for the per-ENI routing tables in the new
ENI datapath. Upcoming commits will use this value and implement the new
datapath.

See https://github.com/cilium/cilium/issues/14336.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
api: Expose egress-multi-home-ip-rule-compat flag

[ upstream commit d5ad26c10; forward-ported from v1.7 tree ]

This is important for use in the CNI to decide whether to use the new
ENI datapath (see previous commit for context) or the original datapath.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
routing: Refactor helper to run function in netns

[ upstream commit 18da31432; forward-ported from v1.7 tree ]

This makes it usable for an upcoming commit which adds a new test suite
to this package.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
routing: Remove unnecessary debug logs from test

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
datapath, cni: Refactor masquerade parameter

Previously, we required `masquerade` to be passed in as a parameter in
multiple places, rather than passing it in once and storing it in
`RoutingInfo`. This commit refactors the code to do as such.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon: Auto-set egress-multi-home-ip-rule-compat for Azure

This is required because the Azure IPAM datapath does not yet support
the new scheme that the ENI IPAM datapath now supports (see earlier
commits). The motivation to make it impossible for Azure users to
accidentally (or otherwise) set this flag to false, which would break
the Azure datapath.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

twpayne reviewed (approved):
LGTM. There are a couple of items from the 1.7 PR review (#14337) that it would be nice to see at some point:
* Use more specific terms than "new" and "old" schemes
* Add a sanity check that the generated interface numbers do not conflict with any reserved numbers

vadorovsky commented on pkg/datapath/linux/routing/info.go:
```diff
@@ -77,9 +86,16 @@ func parse(gateway string, cidrs []string, macAddr string, masquerade bool) (*Ro
 		return nil, fmt.Errorf("invalid mac: %s", macAddr)
 	}
 
+	parsedIfaceNum, err := strconv.ParseInt(ifaceNum, 10, 64)
```
I'd suggest to do:

```suggestion
	parsedIfaceNum, err := strconv.Atoi(ifaceNum)
```

To parse it as `int` from the beginning, instead of parsing as `int64` and then converting to `int`. I think that should fix the warning below.

vadorovsky commented on pkg/datapath/linux/routing/info.go:
```diff
@@ -77,9 +86,16 @@ func parse(gateway string, cidrs []string, macAddr string, masquerade bool) (*Ro
+		IPv4CIDRs:       parsedCIDRs,
+		MasterIfMAC:     parsedMAC,
+		Masquerade:      masquerade,
+		InterfaceNumber: int(parsedIfaceNum),
```
and then

```suggestion
		InterfaceNumber: parsedIfaceNum,
```

vadorovsky reviewed (commented):
Just a minor nit. Everything else looks good to me.

vadorovsky reviewed (approved):
LGTM, thanks!

joestringer reviewed (approved):
I reviewed the original PR and was happy with it, and I've now reviewed the commits that were not in the original PR. Those LGTM.

Let me know if you think I should pay closer attention to any of the remaining commits.
